transport/swift/swift.go

```diff
@@ -0,0 +1,107 @@
+package swift
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/imgproxy/imgproxy/v3/config"
+	"github.com/ncw/swift/v2"
+)
+
+type transport struct {
+	con *swift.Connection
+}
+
+func New() (http.RoundTripper, error) {
+	c := &swift.Connection{
+		UserName:       config.SwiftUsername,
+		ApiKey:         config.SwiftAPIKey,
+		AuthUrl:        config.SwiftAuthURL,
+		AuthVersion:    config.SwiftAuthVersion,
+		Domain:         config.SwiftDomain, // v3 auth only
+		Tenant:         config.SwiftTenant, // v2 auth only
+		Timeout:        time.Duration(config.SwiftTimeoutSeconds) * time.Second,
+		ConnectTimeout: time.Duration(config.SwiftConnectTimeoutSeconds) * time.Second,
+	}
+
+	ctx := context.Background()
+
+	err := c.Authenticate(ctx)
+
+	if err != nil {
+		return nil, fmt.Errorf("swift authentication error: %s", err)
+	}
+
+	return transport{con: c}, nil
+}
+
+func (t transport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	container, path, err := parseObjectURL(req.URL.Path)
```

Comment:
Let's make it `swift://container/path`, not `swift:///container/path`.

```suggestion
	container := req.URL.Host
	path := req.URL.Path
```

You can trim the leading slash with `strings.trimPrefix(req.URL.Path, "/")` if this is required by Swift.

Reply:
That was a bad commit from my test code. Apologies! Will fix that.

Reply:
Sounds good.

A standard OpenStack Swift deployment has the resource path in the format of `/v1/{account}/{container}/{object}` according to https://docs.openstack.org/swift/latest/api/object_api_v1_overview.html . An example Swift object url may look like this `http://127.0.0.1:8080/v1/AUTH_test/test/foo/test.png` , i.e.

```
               Account: AUTH_test
             Container: test
                Object: foo/test.png
```

It should be quite trivial for the user to convert that to `swift://test/foo/test.png` before passing it to imgproxy. I'll document that.

Reply:
That's implemented now with documentation update. (I just realised you suggested the same thing in the previous comment. Apologies for missing that.) Please have another look.

Reply:
I rebased btw. Please check https://github.com/imgproxy/imgproxy/pull/837/commits/675f7bf1f99c9f5b2ac958c69568fe9d8b46c55e .